packages/snaps-rpc-methods/src/permitted/setState.ts

```diff
@@ -0,0 +1,222 @@
+  /**
+   * Check if the requesting origin has a given permission.
+   *
+   * @param permissionName - The name of the permission to check.
+   * @returns Whether the origin has the permission.
+   */
+  hasPermission: (permissionName: string) => boolean;
+
+  /**
+   * Get the state of the requesting Snap.
+   *
+   * @param snapId - The ID of the Snap.
+   * @param encrypted - Whether the state is encrypted.
+   * @returns The current state of the Snap.
+   */
+  getSnapState: (
+    snapId: string,
+    encrypted: boolean,
+  ) => Promise<Record<string, Json>>;
+
+  /**
+   * Wait for the extension to be unlocked.
+   *
+   * @returns A promise that resolves once the extension is unlocked.
+   */
+  getUnlockPromise: (shouldShowUnlockRequest: boolean) => Promise<void>;
+
+  /**
+   * Update the state of the requesting Snap.
+   *
+   * @param snapId - The ID of the Snap.
+   * @param newState - The new state of the Snap.
+   * @param encrypted - Whether the state should be encrypted.
+   */
+  updateSnapState: (
+    snapId: string,
+    newState: Record<string, Json>,
+    encrypted: boolean,
+  ) => Promise<void>;
+};
+
+const SetStateParametersStruct = objectStruct({
+  key: optional(string()),
+  value: JsonStruct,
+  encrypted: optional(boolean()),
+});
+
+export type SetStateParameters = InferMatching<
+  typeof SetStateParametersStruct,
+  SetStateParams
+>;
+
+/**
+ * The `snap_setState` method implementation.
+ *
+ * @param request - The JSON-RPC request object.
+ * @param response - The JSON-RPC response object.
+ * @param _next - The `json-rpc-engine` "next" callback. Not used by this
+ * function.
+ * @param end - The `json-rpc-engine` "end" callback.
+ * @param hooks - The RPC method hooks.
+ * @param hooks.hasPermission - Check whether a given origin has a given
+ * permission.
+ * @param hooks.getSnapState - Get the state of the requesting Snap.
+ * @param hooks.getUnlockPromise - Wait for the extension to be unlocked.
+ * @param hooks.updateSnapState - Update the state of the requesting Snap.
+ * @returns Nothing.
+ */
+async function setStateImplementation(
+  request: JsonRpcRequest<SetStateParameters>,
+  response: PendingJsonRpcResponse<SetStateResult>,
+  _next: unknown,
+  end: JsonRpcEngineEndCallback,
+  {
+    hasPermission,
+    getSnapState,
+    getUnlockPromise,
+    updateSnapState,
+  }: SetStateHooks,
+): Promise<void> {
+  const { params } = request;
+
+  if (!hasPermission('snap_manageState')) {
+    return end(providerErrors.unauthorized());
+  }
+
+  try {
+    const validatedParams = getValidatedParams(params);
+    const { key, value, encrypted = true } = validatedParams;
+
+    if (encrypted) {
+      await getUnlockPromise(true);
+    }
+
+    // We expect the MM middleware stack to always add the origin to requests
+    const { origin } = request as JsonRpcRequest & { origin: string };
+    const state = await getSnapState(origin, encrypted);
+
+    const newState = set(state, key, value);
```

Comment:
We may wanna run some benchmarking on this to see if we can make it faster

Reply:
This implementation should be pretty optimised already (using for-loop instead of `reduce` etc.), but if you have suggestions for further (possible) optimisations, I'm happy to benchmark it.